### PR TITLE
add delete limit to mysql

### PR DESCRIPTION
**Pull request by Polandia94**
add limit on delete on mysql to resolve issue in:
https://github.com/sqlalchemy/sqlalchemy/issues/11764

### CODE REVIEW

File: test/orm/dml/test_update_delete_where.py [hunk 1]
```diff
@@ -2585,6 +2585,27 @@ def test_update_from_multitable_same_names(self):
...
         )
 
+    def test_delete_args(self):
+        Data = self.classes.Data
+        session = fixture_session()
+        delete_args = {"mysql_limit": 1}
+
+        m1 = testing.mock.Mock()
+
+        @event.listens_for(session, "after_bulk_update")
```
Comment (zzzeek) [inline test/orm/dml/test_update_delete_where.py:2595]: after_bulk_delete ?

File: test/orm/dml/test_update_delete_where.py [hunk 2]
```diff
@@ -2585,6 +2585,27 @@ def test_update_from_multitable_same_names(self):
             ],
         )
 
+    def test_delete_args(self):
```
Comment (zzzeek) [inline test/orm/dml/test_update_delete_where.py:2588]: this test should probably be right underneath test_update_args in ExpressionUpdateTest

File: lib/sqlalchemy/sql/compiler.py [hunk 3]
```diff
@@ -6102,6 +6102,10 @@ def update_limit_clause(self, update_stmt):
         """Provide a hook for MySQL to add LIMIT to the UPDATE"""
         return None
 
+    def delete_limit_clause(self, delete_stmt):
```
Comment (zzzeek) [inline lib/sqlalchemy/sql/compiler.py:6105]: yup....looks like we did update_limit_clause the same way, so this is consistent....

File: lib/sqlalchemy/dialects/mysql/base.py [hunk 4]
```diff
@@ -473,6 +473,10 @@ def connect(dbapi_connection, connection_record):
 
     update(..., mysql_limit=10, mariadb_limit=10)
 
+* DELETE with LIMIT::
+
+    delete(..., mysql_limit=10, mariadb_limit=10)
```
Comment (zzzeek) [inline lib/sqlalchemy/dialects/mysql/base.py:478]: would need a `.. versionadded::` in docs

**Review by zzzeek — CHANGES_REQUESTED** [review]
hey there -

this looks pretty good.

it would be most straightforward to target this towards release 2.1.    a 2.0 backport would be possible but easier if we can just keep it on 2.1 since this issue has been open a long time anyway.

some questions on the tests

**Comment by Polandia94** [comment]
> it would be most straightforward to target this towards release 2.1. a 2.0 backport would be possible but easier if we can just keep it on 2.1 since this issue has been open a long time anyway.

I'm not sure how to do this. I added .. versionadded:: 2.1 to the docs, is there a separate branch of 2.1 to point this PR to that branch?

> some questions on the tests

Comments solved.

**Comment by zzzeek** [comment]
> > it would be most straightforward to target this towards release 2.1. a 2.0 backport would be possible but easier if we can just keep it on 2.1 since this issue has been open a long time anyway.
> 
> I'm not sure how to do this. I added .. versionadded:: 2.1 to the docs, is there a separate branch of 2.1 to point this PR to that branch?

dont worry about that part , as long as this is on main it's fine

> 
> > some questions on the tests
> 
> Comments solved.

thanks!

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 7db73666185e0c50b21a31e3b436fb5bc9b9f44b of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change 7db73666185e0c50b21a31e3b436fb5bc9b9f44b: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5607

**Comment by Polandia94** [comment]
I did a change on base and on tests. I runned tox against mysql and pass the test. @zzzeek

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision b8c7e0d245d04d95736decef8db59b7a6f900b6c of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset b8c7e0d245d04d95736decef8db59b7a6f900b6c added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5607

File: lib/sqlalchemy/dialects/mysql/base.py (hunk 4 above)
Comment (CaselIT) [reply lib/sqlalchemy/dialects/mysql/base.py:480]: this needs to be after the delete, not before

File: lib/sqlalchemy/orm/query.py [hunk 5]
```diff
@@ -3169,7 +3177,9 @@ def delete(
 
         """
 
-        bulk_del = BulkDelete(self)
+        delete_args = delete_args or {}
```
Comment (CaselIT) [inline lib/sqlalchemy/orm/query.py:3180]: bulk delete accepts none, so let's avoid creating a dict here, and just pass none to it

**Review by CaselIT — CHANGES_REQUESTED** [review]
Could you also add a changelog?
It should be placed in this folder https://github.com/sqlalchemy/sqlalchemy/tree/main/doc/build/changelog/unreleased_21, where you also find some examples. The number is refers to the issue

**Comment by Polandia94** [comment]
@CaselIT Changelog added. Run tox and locally is passing the tests

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision e34708374c67e016cda88919109fec5e6462eced of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset e34708374c67e016cda88919109fec5e6462eced added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5607

**Comment by sqla-tester** [comment]
**Michael Bayer** (zzzeek) wrote:

@caselit I have done a backport here for 2.0 considering there were issues with the update() construct as well (wrong docs, no checking for int value)

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5607

**Comment by sqla-tester** [comment]
**Michael Bayer** (zzzeek) wrote:

we have tests for the caching  https://github.com/sqlalchemy/sqlalchemy/blob/f1335227c0f5da63f9a4f9179bdb8ef6916bc758/test/sql/test_compare.py#L667

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5607

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5607 has been **merged**. Congratulations! :)

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5627 has been **merged**. Congratulations! :)

**Pull request closed without merging**